debezium-server/debezium-server-ybexporter/src/main/java/io/debezium/server/ybexporter/StreamingWriterJson.java

```diff
@@ -5,87 +5,137 @@
-    private ObjectWriter ow;
-    private FileOutputStream fos;
-    private FileDescriptor fd;
+    private QueueSegment currentQueueSegment;
+    private long currentQueueSegmentIndex = 0;
+
 
     public StreamingWriterJson(String datadirStr) {
         dataDir = datadirStr;
 
-        var fileName = String.format("%s/%s", dataDir, QUEUE_FILE_NAME);
-        try {
-            fos = new FileOutputStream(fileName, true);
-            fd = fos.getFD();
-            var f = new FileWriter(fd);
-            writer = new BufferedWriter(f);
-
+        // mkdir cdc
+        File queueDir = new File(String.format("%s/%s", dataDir, QUEUE_FILE_DIR));
+        if (!queueDir.exists()){
+            boolean dirCreated = queueDir.mkdir();
+            if (!dirCreated){
+                throw new RuntimeException("failed to create dir for cdc");
+            }
         }
-        catch (IOException e) {
-            throw new RuntimeException(e);
+        recoverStateFromDisk();
+        if (currentQueueSegment == null){
+            createNewQueueSegment();
         }
-        ow = new ObjectMapper().writer().withDefaultPrettyPrinter();
     }
 
-    @Override
-    public void writeRecord(Record r) {
+    private void createNewQueueSegment(){
+        currentQueueSegment = new QueueSegment(getFilePathWithIndex(currentQueueSegmentIndex));
+    }
+
+    private void recoverStateFromDisk(){
+        recoverLatestQueueSegment();
+    }
+
+    /**
+     * This function reads the /cdc dir for all the queue segment files.
+     * Then it retrieves the index number from the paths, and then finds
+     * the max index - which is the latest queue segment that was written to.
+     * If no files are found, we just return.
+     */
+    private void recoverLatestQueueSegment(){
+        // read dir to find all queue files
+        Path queueDirPath = Path.of(dataDir, QUEUE_FILE_DIR);
+        String searchGlob = String.format("%s.*.%s", QUEUE_FILE_NAME, QUEUE_FILE_EXTENSION);
+        ArrayList<Path> filePaths = new ArrayList<>();
         try {
-            String cdcJson = ow.writeValueAsString(generateCdcMessageForRecord(r));
-            writer.write(cdcJson);
-            writer.write("\n");
-            LOGGER.info("Writing CDC message = {}", cdcJson);
+            DirectoryStream<Path> stream = Files.newDirectoryStream(queueDirPath, searchGlob);
+            for (Path entry: stream) {
+                filePaths.add(entry);
+            }
+            if (filePaths.size() == 0){
+                // no files found. nothing to recover.
+                LOGGER.info("No files found matching {}. Nothing to recover", searchGlob);
+                return;
+            }
+            // extract max index of all files
+            int maxIndex = 0;
+            for(Path p: filePaths){
+                // get the substring after the last occurence of "." and convert to ind
+                String pathWithoutExtention = p.toString().replace("."+QUEUE_FILE_EXTENSION, "");
```

Comment:
I used a different logic for extracting segment number:
```
fileName = filepath.Base(p.toString())
segmentNum = fileName[len("queue."): len(fileName) - len(".ndjson")]
```

This logic also looks fine to me. We can go with same logic at both places